sui_core/src/gateway_state.rs

```diff
@@ -798,32 +530,28 @@ where
         gas_payment: ObjectID,
         recipient: SuiAddress,
     ) -> Result<TransactionData, anyhow::Error> {
-        let account = self.get_or_create_account(signer)?;
-        let object_ref = account.latest_object_ref(&object_id)?;
-        self.get_object_info(object_id)
-            .await?
-            .object()?
-            .is_transfer_eligible()?;
+        // TODO: We should be passing in object_ref directly instead of object_id.
+        let object = self.get_object(&object_id).await?;
+        let object_ref = object.compute_object_reference();
+        let gas_payment = self.get_object(&gas_payment).await?;
+        let gas_payment_ref = gas_payment.compute_object_reference();
 
-        let account = self.get_or_create_account(signer)?;
-        let gas_payment = account.latest_object_ref(&gas_payment)?;
-        let data = TransactionData::new_transfer(recipient, object_ref, signer, gas_payment);
+        let data = TransactionData::new_transfer(recipient, object_ref, signer, gas_payment_ref);
 
         Ok(data)
     }
 
-    async fn sync_account_state(&mut self, account_addr: SuiAddress) -> Result<(), anyhow::Error> {
-        self.try_complete_pending_transactions(account_addr).await?;
-
+    // TODO: Get rid of the sync API.
+    // https://github.com/MystenLabs/sui/issues/1045
+    async fn sync_account_state(&self, account_addr: SuiAddress) -> Result<(), anyhow::Error> {
```

Comment:
I think that here wht we would need is an time-index of changes for objects with a specific owner, no?